### PR TITLE
[MINI-3131] Feature/update get access token error type

### PR DESCRIPTION
**Pull request by munir-rakuten**
# Description
Added Specific error types to use from host app to show in mini apps when failed to fetch access token.
Demo App Can send 

MiniAppAccessTokenError.scopesNotSupportedError
MiniAppAccessTokenError.audienceNotSupportedError
MiniAppAccessTokenError.authorizationFailureError(message = "authorization failure error msg")
MiniAppAccessTokenError.custom(message = "custom error msg")

## Links
MINI-3131

# Checklist
- [x] I have read the [contributing guidelines](../.github/CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](../.github/CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [x] I wrote unit test for new code
- [x] I removed all sensitive data before every commit, including API endpoints and keys
- [x] I ran `./gradlew check` without errors

### CODE REVIEW

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppErrors.kt
Comment (minh-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppErrors.kt:8]: The spec of `Authorization failure` declared that it is a custom error message sent from hostapp.
If we go only with `AccessTokenErrorType.errorKey` then the hostapp user cannot pass the error message so

we can create a `MiniAppAccessTokenError` class which contains `error: AccessTokenErrorType = OtherError` and `message: String = ""`.

When user does not pass the `message` which is empty as default, we will let the errorCallback send `AccessTokenErrorType.errorKey`.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt
Comment (munir-rakuten) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt:106]: @minh-rakuten  yes i wanted to do like this..bt it will be a major change and can break the previous implementation...

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt
Comment (khairul-alam-licon) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt:117]: I thought, just returning `AccessTokenErrorType.getValue(errorType.error)` would be enough 🤔 
Or, I am not sure about the context of `AuthorizationFailureError` here.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt
Comment (khairul-alam-licon) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt:198]: Then, we need to update the user-guide too:
Screen Shot 2021-05-25 at 0 03 10

File: CHANGELOG.md
Comment (khairul-alam-licon) [inline CHANGELOG.md:9]: I think, if we still keep the old one, then we can replace the word "Updated" here by "Added" with mentioning that the old one is being deprecated.

**Review by khairul-alam-licon — COMMENTED** [review]
Looking really close! 
I have left few concerns.

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt
Comment (munir-rakuten) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt:198]: yes we need to update it @khairul-alam-licon

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt
Comment (munir-rakuten) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt:117]: for the AuthorizationFailureError and Error type the message will be set by the host app so it needs the case @khairul-alam-licon

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt
Comment (munir-rakuten) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt:104]: thanks, yes it can be done like that, @corycaywood  may suggest, should we support it?

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridgeDispatcher.kt
Comment (munir-rakuten) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridgeDispatcher.kt:68]: yes, no need, bt i left it just for clearity... @khairul-alam-licon

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt:13]: I think both of these should be nullable.
@Climbatize Can the JS SDK handle the case where one or both is null?

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt:11]: Maybe this type should be something like `MiniAppBridgeError` so it's more clear where it's used. What do you think?

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt:15]: I think this constructor isn't needed for anything, right?

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppErrorTypes.kt
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppErrorTypes.kt:6]: I'm not sure if these need to be defined as an enum. Shouldn't it be enough to define them as just strings since the `MiniAppError` accepts a string as the `type`?

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt
Comment (corycaywood) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt:11]: It seems like `MiniAppError` should be internal as this is exposing the JSON format that we're using. Also, we want them to only provide the supported types of errors to `getAccessToken`, so I think we need at least three classes:

- `MiniAppBridgeErrorModel` - internal data model (same as `MiniAppError` here currently)
- `MiniAppBridgeError` - abstract public class that's used as base class for errors to pass to mini app
- `MiniAppAccessTokenError` - public class that extends `MiniAppBridgeError` and is used as the error type for `getAccessToken`. It can also have pre-defined errors in the `companion object` to construct the specific errors. (i.e. `MiniAppAccessTokenError.audienceNotSupportedError`)

Do you think that makes sense?

**Review by corycaywood — COMMENTED** [review]
Looking good, but I think we need a few tweaks for the public interface.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt
Comment (munir-rakuten) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt:11]: @corycaywood  yes, we can use `MiniAppBridgeError` instead of `MiniAppError`, May be JS SDK use `MiniAppError`, so i thought it may look consistent in both platform, i will update it no problem

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt
Comment (munir-rakuten) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt:15]: yes, i will remove this constructor.. @corycaywood

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt
Comment (Climbatize) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/errors/MiniAppError.kt]: @corycaywood I managed this case so that's even fine to send `{}`

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt
Comment (Climbatize) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt:104]: @khairul-alam-licon @munir-rakuten I think we don't have to support the previous implementation there, as the callback is generated by the SDK and not by the host app, and is managed by the JS bridge

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt
Comment (Climbatize) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/display/MiniAppDisplayActivity.kt:198]: I think you still need to update `USERGUIDE.md` @munir-rakuten :)

**Review by Climbatize — COMMENTED** [review]
LGTM. I'll approve the request when the JS-Miniapp PR https://github.com/rakutentech/js-miniapp/pull/130 will be validated and the submodule upgraded there, and the user guide updated 👍

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt
Comment (Climbatize) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/userinfo/UserInfoBridge.kt:111]: this has to be done on line 115 where you just send a regular `String` as an error and not a `MiniAppBridgeErrorModel`

I think you will have to modify `doesAccessTokenMatch` method to return the right error type and not just a Boolean

**Review by Climbatize — CHANGES_REQUESTED** [review]

**Review by khairul-alam-licon — APPROVED** [review]
Looks good 👍 
Also, need to update with JS side before merging this PR.

**Review by Climbatize — APPROVED** [review]

**Pull request merged**
